### PR TITLE
feat: Added Smart Mode that Automatically detects changed Environment…

### PR DESCRIPTION
**Pull request by fritzduchardt**
…s and Applications

### CODE REVIEW

File: internal/myks/git.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+func extractChangedFilePathsWithoutStatus(cfs []ChangedFile, status string) []string {
+	var paths []string
+	for _, cf := range cfs {
+		if status == "" || cf.status != status {
+			paths = append(paths, cf.path)
+		}
+	}
+	return paths
+}
```
Comment (Zebradil) [inline internal/myks/git.go:64]: - Checking for `status` can be moved out of the loop.
- I think checking for status should be done outside the function.
- I'd use a more generic function, which accepts a test callback:
```go
func filter[T any](ss []T, test func(T) bool) (ret []T) {
    for _, s := range ss {
        if test(s) {
            ret = append(ret, s)
        }
    }
    return
}
```
(borrowed from [here](https://stackoverflow.com/a/37563128/2227895))

Comment (Zebradil) [reply internal/myks/git.go:64]: But I'm not blocking you with this suggestion, we can refactor this later.

File: internal/myks/globe.go [hunk 2]
```diff
@@ -346,13 +350,15 @@ func (g *Globe) collectEnvironmentsInPath(searchPath string) {
...
 		}
 		if d != nil && d.IsDir() {
-			_, err := os.Stat(filepath.Join(path, g.EnvironmentDataFileName))
-			if err == nil {
-				env := NewEnvironment(g, path)
-				if env != nil {
-					g.environments[path] = env
-				} else {
-					log.Warn().Str("path", path).Msg("Unable to collect environment, skipping")
+			if path != g.EnvironmentBaseDir {
```
Comment (Zebradil) [inline internal/myks/globe.go:353]: Is it to remove the warning from logs?

File: internal/myks/globe.go [hunk 3]
```diff
@@ -151,7 +152,10 @@ func (g *Globe) Init(asyncLevel int, searchPaths []string, applicationNames []st
 
 	dataSchemaFileName := filepath.Join(g.RootDir, g.ServiceDirName, g.TempDirName, g.DataSchemaFileName)
 	if _, err := os.Stat(dataSchemaFileName); err != nil {
-		log.Warn().Err(err).Msg("Unable to find data schema file, creating one")
+		log.Warn().Msg("Unable to find data schema file, creating one")
+		if err := os.MkdirAll(filepath.Dir(dataSchemaFileName), 0o750); err != nil {
+			log.Fatal().Err(err).Msg("Unable to create data schema file directory")
+		}
 		if err := os.WriteFile(dataSchemaFileName, dataSchema, 0o600); err != nil {
```
Comment (Zebradil) [inline internal/myks/globe.go:159]: I think thins won't be necessary after merging from release-2.0.

File: internal/myks/globe.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ type Globe struct {
 
 	// Base directory for environments
 	EnvironmentBaseDir string `default:"envs"`
-	// Prefix for kubernetes namespaces
+	// Prefix for kubernetes namespaces"
```
Comment (Zebradil) [inline internal/myks/globe.go:44]:
```suggestion
	// Prefix for kubernetes namespaces
```

**Review by Zebradil — COMMENTED** [review]
Wow, that's solid!

But could we get a description of this mode? To better understand edge cases and application areas. For example, what if there are several base branches (main←development←feature)? What if I want to apply smart mode locally for all commits on my feature branch (like it's done for CI)?

We will then be able to use it for help and documentation.

File: cmd/root.go [hunk 5]
```diff
@@ -44,7 +45,16 @@ var rootCmd = &cobra.Command{
...
+			// smart mode requires instantiation of globe object to get the list of environments
+			// the globe object will not be used later in the process. It is only used to get the list of all environments and their apps.
+			globeAllEnvsAndApps := myks.New(".")
+			targetEnvironments, targetApplications, err = globeAllEnvsAndApps.InitSmartMode()
+			if err != nil {
+				log.Warn().Err(err).Msg("Unable to run Smart Mode. Rendering everything.")
+			}
+			if targetEnvironments == nil && targetApplications == nil {
+				log.Warn().Err(err).Msg("Smart Mode does not identify changes. Rendering everything.")
+			}
```
[the remark this answers is not in the document]
Comment (fritzduchardt) [reply cmd/root.go:57]: Was thinking about that..let's do it.

File: internal/myks/globe.go (hunk 2 above)
Comment (fritzduchardt) [reply internal/myks/globe.go:353]: Sure, lets set it to debug

**Comment by elyesbenamor** [comment]
@fritzduchardt for auto mode we have also some issue rendering and detecting rendered files
Bug:
- render an application 
-  remove the rendered folder from the environment 
- try to render again to see if it will detect the removed directory 
`1:31PM INF Setting log level to: info
1:31PM INF [global] Smart mode detected changes in environments: [], applications: []
1:31PM WRN Smart Mode did not find any changes. Existing.`
all in all it will not detect the removed directory and it wont render anything

**Comment by fritzduchardt** [comment]
> @fritzduchardt for auto mode we have also some issue rendering and detecting rendered files Bug:
> 
> * render an application
> * remove the rendered folder from the environment
> * try to render again to see if it will detect the removed directory
>   `1:31PM INF Setting log level to: info 1:31PM INF [global] Smart mode detected changes in environments: [], applications: [] 1:31PM WRN Smart Mode did not find any changes. Existing.`
>   all in all it will not detect the removed directory and it wont render anything

Fair point - will look into it

**Comment by kbudde** [comment]
> > @fritzduchardt for auto mode we have also some issue rendering and detecting rendered files Bug:
> > 
> > * render an application
> > * remove the rendered folder from the environment
> > * try to render again to see if it will detect the removed directory
> >   `1:31PM INF Setting log level to: info 1:31PM INF [global] Smart mode detected changes in environments: [], applications: [] 1:31PM WRN Smart Mode did not find any changes. Existing.`
> >   all in all it will not detect the removed directory and it wont render anything
> 
> Fair point - will look into it

Let's not over complicate the smart mode. 
I think this can be solved in #63

**Comment by fritzduchardt** [comment]
> > > @fritzduchardt for auto mode we have also some issue rendering and detecting rendered files Bug:
> > > 
> > > * render an application
> > > * remove the rendered folder from the environment
> > > * try to render again to see if it will detect the removed directory
> > >   `1:31PM INF Setting log level to: info 1:31PM INF [global] Smart mode detected changes in environments: [], applications: [] 1:31PM WRN Smart Mode did not find any changes. Existing.`
> > >   all in all it will not detect the removed directory and it wont render anything
> > 
> > 
> > Fair point - will look into it
> 
> Let's not over complicate the smart mode. I think this can be solved in #63

Agreed - a stab at coding this reveals that the complexity increase would be substantial, especially, because the env structure in the rendered folder is flattened. So, once @Zebradil gives me his well, I am happy to merge this.

**Review by Zebradil — APPROVED** [review]
🚀

**Pull request merged**